### DISCUSSION TITLE
List all buckets (Redis) - monitoring

### DISCUSSION THREAD

**Discussion by PatrykGala** [vladimir-bukhtoyarov/bucket4j#313] [Q&A] [answered]
Hi, 
I would like to list all buckets in lettuce implementation for monitoring purposes. 
How can I get all keys in distributed system? How to read remaining tokens? 

Patryk

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Hello @PatrykGala 

1. Bucket4j does not provide functionality to get all keys(I suppose that letuce too), you should do this job via Redis API.
2. When you know the key of a bucket then remaining tokens can be found like this:
```java
   Optional<BucketConfiguration> optConf =  proxyManager.getProxyConfiguration(key);
   int remainingTokens = optConf.isPresent() ?  proxyManager.builder().build(key, optConf.get()) : 0;
```

**Reply by PatrykGala** [reply]
Thank you, I will try read all by [KEYS](https://redis.io/commands/keys/) Redis command.

In our system we have to limit 50-60k request per second, do you have experience with Redis and bucket4j for similar workloads? We've started to implementing custom library with [INCR](https://redis.io/commands/incr/) command, but first we'd like to test Bucket4j. This library use compare and swap algorithm, is it possible override this to increment command?

**Reply by vladimir-bukhtoyarov** [reply]
1. For CAS based algorithms, the max rate over single key  takes much more effects from scalability point of view than total rate over all keys. You should use [batching optimization](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/RemoteBucketBuilder.java#L63) in case of huge rate on the same key.
2. INCR command can not be used inside bucket4j, because bucket state is stored as array of bytes. There is plan to optimize CAS via replacing it by LUA stored procedures, but it will not be done in 2022.

**Reply by PatrykGala** [reply]
1. We are testing `Optimization.delaying`  and result looks optimistic :) 
2. What is the status of the lua scripts in `experiment` directory? Does it tested and works properly? I can contribute and attach this scripts to java library, but I'm not sure that the script works.

**Reply by vladimir-bukhtoyarov** [reply]
Existed lua scripts are obsolete, they were never working. Activity related with lua will be ressurected next year.

**Reply by PatrykGala** [reply]
Thank you for your support 👍 
Let me know when you will publish snapshot with lua scripts in next year.
